src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -3262,6 +3262,293 @@ void C2_MacroAssembler::stringL_indexof_char(Register str1, Register cnt1, Regis
+    // h = (31 * h) or (h << 5 - h);
+    movl(tmp, result);
+    shll(result, 5);
+    subl(result, tmp);
+    // h += ary1[i];
+    arrays_hashcode_elload(tmp, Address(ary1, index, Address::times(elsize), idx * elsize), eltype, is_string_hashcode);
+    addl(result, tmp);
+  }
+  addl(index, 4);
+  jmpb(SHORT_UNROLLED_LOOP_BEGIN);
+  // }
+  // for (; i < cnt1; i += 1) {
+  bind(SHORT_SCALAR_LOOP_BEGIN);
+  // i < cnt1;
+  cmpl(index, cnt1);
+  jcc(Assembler::greaterEqual, END);
+
+  // h = (31 * h) or (h << 5 - h);
+  movl(tmp, result);
+  shll(result, 5);
+  subl(result, tmp);
+  // h += ary1[i];
+  arrays_hashcode_elload(tmp, Address(ary1, index, Address::times(elsize)), eltype, is_string_hashcode);
+  addl(result, tmp);
+  // i += 1;
+  incrementl(index);
+  jmpb(SHORT_SCALAR_LOOP_BEGIN);
+  // }
+
+  if (generate_vectorized_loop) {
+  // } else { // cnt1 >= 32
+  address power_of_31_backwards = pc();
+  emit_int32( 2111290369);
+  emit_int32(-2010103841);
+  emit_int32(  350799937);
+  emit_int32(   11316127);
+  emit_int32(  693101697);
+  emit_int32( -254736545);
+  emit_int32(  961614017);
+  emit_int32(   31019807);
+  emit_int32(-2077209343);
+  emit_int32(  -67006753);
+  emit_int32( 1244764481);
+  emit_int32(-2038056289);
+  emit_int32(  211350913);
+  emit_int32( -408824225);
+  emit_int32( -844471871);
+  emit_int32( -997072353);
+  emit_int32( 1353309697);
+  emit_int32( -510534177);
+  emit_int32( 1507551809);
+  emit_int32( -505558625);
+  emit_int32( -293403007);
+  emit_int32(  129082719);
+  emit_int32(-1796951359);
+  emit_int32( -196513505);
+  emit_int32(-1807454463);
+  emit_int32( 1742810335);
+  emit_int32(  887503681);
+  emit_int32(   28629151);
+  emit_int32(     923521);
+  emit_int32(      29791);
+  emit_int32(        961);
+  emit_int32(         31);
+  emit_int32(          1);
+
+  bind(LONG);
+
+  // int coef = 1;
+  movl(coef, 1);
+  // int i = cnt1 - 1;
+  movl(index, cnt1);
+  decrementl(index);
+  // bound = cnt1 & ~(32-1);
+  bound = cnt1;
+  movl(bound, cnt1);
+  andl(bound, ~(32-1));
+
+  // result = 0;
+  if (!is_string_hashcode) {
+    movl(result, 0);
+  }
+
+  // for (; i >= bound; i -= 1) {
+  bind(LONG_SCALAR_LOOP_BEGIN);
+  // i >= bound;
+  cmpl(index, bound);
+  jccb(Assembler::less, LONG_SCALAR_LOOP_END);
+  // result += coef * ary1[i];
+  arrays_hashcode_elload(tmp, Address(ary1, index, Address::times(elsize)), eltype, is_string_hashcode);
+  imull(tmp, coef);
+  addl(result, tmp);
+  // coef *= 31;
+  movl(tmp, 31);
+  imull(coef, tmp);
+  // i -= 1;
+  decrementl(index);
+  jmpb(LONG_SCALAR_LOOP_BEGIN);
+  bind(LONG_SCALAR_LOOP_END);
+  // }
+
+  for (int idx = 0; idx < 4; idx++) {
+    // vresult = IntVector.zero(I256);
+    vpxor(vresult[idx], vresult[idx]);
+  }
+  // vnext = IntVector.broadcast(I256, power_of_31_backwards[0]);
+  movdl(vnext, InternalAddress(power_of_31_backwards + (0 * sizeof(jint))));
+  vpbroadcastd(vnext, vnext, Assembler::AVX_256bit);
```

Comment:
`vpbroadcastd` can take an `Address` argument instead.

Reply:
An `InternalAddress` isn't an `Address` but an `AddressLiteral`. You can however do `as_Address(InternalAddress(power_of_31_backwards + (0 * sizeof(jint))))`